weave/chobj/chobj.py

```diff
@@ -0,0 +1,1271 @@
+        raise NotImplementedError
+
+    def with_item(self, item_id: uuid.UUID, item_version: uuid.UUID) -> "Ref":
+        raise NotImplementedError
+
+
+@dataclasses.dataclass
+class TableRef(Ref):
+    table_id: uuid.UUID
+
+    def uri(self) -> str:
+        return f"table:///{self.table_id}"
+
+
+@dataclasses.dataclass
+class ValRef(Ref):
+    val_id: uuid.UUID
+    extra: list[str] = dataclasses.field(default_factory=list)
+
+    def uri(self) -> str:
+        u = f"val:///{self.val_id}"
+        if self.extra:
+            u += "/" + "/".join(self.extra)
+        return u
+
+    def with_key(self, key) -> "ValRef":
+        return ValRef(self.val_id, self.extra + ["key", key])
+
+    def with_attr(self, attr) -> "ValRef":
+        return ValRef(self.val_id, self.extra + ["attr", attr])
+
+    def with_index(self, index) -> "ValRef":
+        return ValRef(self.val_id, self.extra + ["index", str(index)])
+
+    def with_item(self, item_id, item_version) -> "ValRef":
+        return ValRef(self.val_id, self.extra + ["id", f"{item_id},{item_version}"])
+
+
+@dataclasses.dataclass
+class ObjectRef(Ref):
+    name: str
+    val_id: uuid.UUID
+    extra: list[str] = dataclasses.field(default_factory=list)
+
+    def uri(self) -> str:
+        u = f"object:///{self.name}/{self.val_id}"
+        if self.extra:
+            u += "/" + "/".join(self.extra)
+        return u
+
+    def with_key(self, key) -> "ObjectRef":
+        return ObjectRef(self.name, self.val_id, self.extra + ["key", key])
+
+    def with_attr(self, attr) -> "ObjectRef":
+        return ObjectRef(self.name, self.val_id, self.extra + ["attr", attr])
+
+    def with_index(self, index) -> "ObjectRef":
+        return ObjectRef(self.name, self.val_id, self.extra + ["index", str(index)])
+
+    def with_item(self, item_id, item_version) -> "ObjectRef":
+        return ObjectRef(
+            self.name, self.val_id, self.extra + ["id", f"{item_id},{item_version}"]
+        )
+
+
+def dataclasses_asdict_one_level(obj):
+    # dataclasses.asdict is recursive. We don't want that when json encoding
+    return {f.name: getattr(obj, f.name) for f in dataclasses.fields(obj)}
+
+
+def pydantic_asdict_one_level(obj: pydantic.BaseModel):
+    return {k: getattr(obj, k) for k in obj.model_fields}
+
+
+class RefEncoder(json.JSONEncoder):
+    def default(self, o):
+        if isinstance(o, uuid.UUID):
+            return {"_type": "UUID", "uuid": o.hex}
+        elif dataclasses.is_dataclass(o):
+            data = dataclasses_asdict_one_level(o)
+            return {"_type": o.__class__.__name__, **data}
+        elif isinstance(o, pydantic.BaseModel):
+            data = pydantic_asdict_one_level(o)
+            return {"_type": o.__class__.__name__, **data}
+        elif isinstance(o, ObjectRecord):
+            return o.__dict__
+        return json.JSONEncoder.default(self, o)
+
+
+def json_dumps(val):
+    return json.dumps(val, cls=RefEncoder)
+
+
+def get_type(val):
```

Comment:
so type here is now just a type name, no longer a weave type? Is it possible to get the weave type of a ref?